### PR TITLE
Cayley Rot3 fixes

### LINKED ISSUE #591: Numerous unit test failures with GTSAM_ROT3_EXPMAP=OFF
## Description

Numerous unit test failures with `GTSAM_ROT3_EXPMAP=OFF`. This configuration apparently is not be exercised by CI.

```
The following tests FAILED:
	 25 - testEssentialMatrix (Failed)
	100 - testAdaptAutoDiff (Failed)
	189 - testSimilarity3 (Failed)
	219 - testPoseBetweenFactor (Failed)
	233 - testPosePriorFactor (Failed)
```

## Steps to reproduce

1. Check out latest develop 
2. Set `GTSAM_ROT3_EXPMAP=OFF`, and also `GTSAM_BUILD_WITH_MARCH_NATIVE=OFF` (separate issue caused by `march=native` reported in https://github.com/borglab/gtsam/issues/590)
3. `make check`

## Expected behavior

Tests pass, and CI should probably exercise this configuration.

## Environment

Ubuntu 16.04, gcc 5.4.0
Mac OS, Xcode 12.1, Apple clang 12.0.0.12000032

## Additional information

Detailed failures on Ubuntu:

```
 25/243 Test  #25: testEssentialMatrix ....................***Failed    0.01 sec
not equal:
expected[0.1; 0.2; 0.3; 0; 0];
actual[0.101183235; 0.202366469; 0.303549704; 0; 0];
/home/cbeall/git/gtsam/gtsam/geometry/tests/testEssentialMatrix.cpp:83: Failure: "assert_equal(d.head(5), actual2, 1e-8)" 
There were 1 failures
```

```
100/243 Test #100: testAdaptAutoDiff ......................***Failed    0.01 sec
/home/cbeall/git/gtsam/gtsam/nonlinear/tests/testAdaptAutoDiff.cpp:189: Failure: "equal_with_abs_tol(expectedP, P)" 
Not equal:
expected:
[
	1.31247;
	1.20573
]
actual:
[
	1.31841;
	1.20368
]
/home/cbeall/git/gtsam/gtsam/nonlinear/tests/testAdaptAutoDiff.cpp:202: Failure: "assert_equal(expectedMeasurement, actual, 1e-6)" 
Not equal:
expected:
[
	1.31247;
	1.20573
]
actual:
[
	1.31841;
	1.20368
]
/home/cbeall/git/gtsam/gtsam/nonlinear/tests/testAdaptAutoDiff.cpp:228: Failure: "assert_equal(expectedMeasurement, actual, 1e-6)" 
Not equal:
expected:
[
	1.31247;
	1.20573
]
actual:
[
	1.31841;
	1.20368
]
/home/cbeall/git/gtsam/gtsam/nonlinear/tests/testAdaptAutoDiff.cpp:234: Failure: "assert_equal(expectedMeasurement, actual2, 1e-6)" 
There were 4 failures
```

```
189/243 Test #189: testSimilarity3 ........................***Failed    0.02 sec
/home/cbeall/git/gtsam/gtsam_unstable/geometry/tests/testSimilarity3.cpp:348: Failure: "Exception: 
Indeterminant linear system detected while working near variable
0 (Symbol: 0).

Thrown when a linear system is ill-posed.  The most common cause for this
error is having underconstrained variables.  Mathematically, the system is
underdetermined.  See the GTSAM Doxygen documentation at
http://borg.cc.gatech.edu/ on gtsam::IndeterminantLinearSystemException for
more information." 
There were 1 failures
```

```
219/243 Test #219: testPoseBetweenFactor ..................***Failed    0.01 sec
not equal:
expected[-0.0298395126; 0.0131455995; 0.0969712917; -0.13918755; -0.142346243; -0.0390885321];
actual[-0.0298135268; 0.0131341516; 0.096886844; -0.145701634; -0.134898526; -0.0421026389];
/home/cbeall/git/gtsam/gtsam_unstable/slam/tests/testPoseBetweenFactor.cpp:136: Failure: "assert_equal(expectedError, actualError, 1e-9)" 
not equal:
expected[0.0173371937; 0.0222228304; -0.012504191; 0.0264132886; 0.0052376953; -7.16127036e-05];
actual[0.0173358202; 0.0222210698; -0.0125032004; 0.0263800787; 0.00540285006; 0.000175859556];
/home/cbeall/git/gtsam/gtsam_unstable/slam/tests/testPoseBetweenFactor.cpp:181: Failure: "assert_equal(expectedError, actualError, 1e-9)" 
There were 2 failures
```

```
233/243 Test #233: testPosePriorFactor ....................***Failed    0.01 sec
not equal:
expected[-0.184137862; 0.139419284; -0.158399297; 0.740211734; -1.19821028; 1.00815609];
actual[-0.182948258; 0.13851858; -0.157375975; 0.766913166; -1.22976117; 0.949345561];
/home/cbeall/git/gtsam/gtsam_unstable/slam/tests/testPosePriorFactor.cpp:131: Failure: "assert_equal(expectedError, actualError, 1e-8)" 
not equal:
expected[-0.0227128853; 0.19376511; 0.276418421; 1.49751986; -0.549375791; 0.452761203];
actual[-0.0224998729; 0.191947887; 0.273826035; 1.36483392; -0.754590051; 0.585710674];
/home/cbeall/git/gtsam/gtsam_unstable/slam/tests/testPosePriorFactor.cpp:173: Failure: "assert_equal(expectedError, actualError, 1e-8)" 
There were 2 failures
```

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR is intended to fix #591.

1. Make `POSE3_EXPMAP` and `ROT3_EXPMAP` depend on each other.
2. CayleyChart Local is fully Eigen based, allowing for Eigen's optimization to do it's job.
3. Updated failing Similarity3 test to only run check for Rot3 Expmap.

### CODE REVIEW

File: cmake/HandleGeneralOptions.cmake
Comment (dellaert) [inline cmake/HandleGeneralOptions.cmake:34]: This is a bit problematic as it is done "silently". Add a message?

File: gtsam/geometry/Rot3M.cpp
Comment (dellaert) [inline gtsam/geometry/Rot3M.cpp:179]: Yes if accompanied by benchmark results on effect on timing.

**Review by dellaert — CHANGES_REQUESTED** [review]
Cool, but need to see timing benchmark before approval :-)

**Comment by dellaert** [comment]
Note, search through "timeXXX.cpp" files, might already be there.

File: cmake/HandleGeneralOptions.cmake
Comment (varunagrawal) [reply cmake/HandleGeneralOptions.cmake:34]: Ideally it should be one flag :slightly_frowning_face: but I need to consult you before making that design change.

File: gtsam/geometry/Rot3M.cpp
Comment (varunagrawal) [reply gtsam/geometry/Rot3M.cpp:179]: Old version is way faster `~650 ms` vs `2248 ms` for the fully Eigen based one. Reverting.

Comment (varunagrawal) [reply gtsam/geometry/Rot3M.cpp:179]: I'll instead add more documentation.

**Comment by varunagrawal** [comment]
I added a CI path for the Cayley Transform in the last commit.

**Comment by varunagrawal** [comment]
@chrisbeall good news! The CI for the Cayley transform passes.

**Comment by dellaert** [comment]
Good, now timing sanity-check and we can land this :-)

**Comment by varunagrawal** [comment]
Timing results for Rot3:

Before:

```
Rodriguez formula given axis angle
1121.42 milliseconds
11214.2 nanosecs/call

Rodriguez formula given canonical coordinates
1136.52 milliseconds
11365.2 nanosecs/call

Expmap
1516.92 milliseconds
15169.2 nanosecs/call

Retract
624.292 milliseconds
6242.92 nanosecs/call

Logmap
694.127 milliseconds
6941.27 nanosecs/call

localCoordinates
617.415 milliseconds
6174.15 nanosecs/call

Slow rotation matrix
1172.43 milliseconds
11724.3 nanosecs/call

Fast Rotation matrix
143.142 milliseconds
1431.42 nanosecs/call
```

After:

```
Rodriguez formula given axis angle
1146.84 milliseconds
11468.4 nanosecs/call

Rodriguez formula given canonical coordinates
1187.28 milliseconds
11872.8 nanosecs/call

Expmap
1598.05 milliseconds
15980.5 nanosecs/call

Retract
624.476 milliseconds
6244.76 nanosecs/call

Logmap
715.463 milliseconds
7154.63 nanosecs/call

localCoordinates
862.56 milliseconds
8625.6 nanosecs/call

Slow rotation matrix
1200.18 milliseconds
12001.8 nanosecs/call

Fast Rotation matrix
144.373 milliseconds
1443.73 nanosecs/call
```

There is negligible impact due to the determinant impact.

**Comment by dellaert** [comment]
I was not worried about the determinant check. I was worried about replacing hand optimized code with much nicer but more high-level code :-) could you just point out the isolated results for Cayley, which is affected by that piece of code? And make sure the right flags are set so Cayley is in fact called ? (You did not mention flags in your comment)

**Comment by varunagrawal** [comment]
I undid the change from hand optimized to high-level, so it is back to hand optimized code now (albeit with an extra determinant check). The flags being set are in the CI and I already verified that. :slightly_smiling_face:

**Comment by dellaert** [comment]
Wait! I liked your nicer code, *if* it is the same speed. What was the timing on that?

**Comment by dellaert** [comment]
Oh, wait, I did not see your comment above. OK, I’ll approve this version, then :-)

**Review by dellaert — APPROVED** [review]

**Pull request merged**
